feat(auth): Add bulk get/delete methods

rsgowman opened the pull request:
This PR allows callers to retrieve a list of users by unique identifier (uid, email, phone, federated provider uid) as well as to delete a list of users.

RELEASE NOTE: Added `getUsers()` and `deleteUsers()` APIs for retrieving and deleting user accounts in bulk.

hiranya911 commented on src/utils/error.ts:
```diff
@@ -649,6 +649,10 @@ export class AuthClientErrorCode {
   };
+  public static USER_NOT_DISABLED = {
+    code: 'user-not-disabled',
+    message: 'The user must be disabled in order to bulk delete it (or you must pass force=true).',
```
Didn't we decide against the `force=true` option in this API?

bojeil-google replied:
I am confused about this. This was approved in https://docs.google.com/document/d/1mG0ilmh8H-C6DYXA9oOQUro2MP0on_9k7Qf4oCLRyBk/edit#

hiranya911 commented on src/auth/auth.ts:
```diff
@@ -263,6 +339,46 @@ export class BaseAuth<T extends AbstractAuthRequestHandler> {
+          successCount: uids.length,
+          errors: [],
+        };
+
```
Nit: Remove the large if block with:

```
if (!validator.isNonEmptyArray(batchDeleteAccountsResponse.errors)) {
  return result;
}
```

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -463,6 +490,45 @@ export const FIREBASE_AUTH_DELETE_ACCOUNT = new ApiSettings('/accounts:delete',
+interface BatchDeleteAccountsRequest {
+  localIds?: string[];
+  force?: boolean;
+}
```
Separate with whitespace.

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -463,6 +490,45 @@ export const FIREBASE_AUTH_DELETE_ACCOUNT = new ApiSettings('/accounts:delete',
+  index?: number;
+  localId?: string;
+  message?: string;
+}
```
Here too

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -452,6 +464,21 @@ export const FIREBASE_AUTH_GET_ACCOUNT_INFO = new ApiSettings('/accounts:lookup'
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INTERNAL_ERROR,
+        'INTERNAL ASSERT FAILED: Server request is missing user identifier');
+    }
```
Should we also add a response validator?

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -463,6 +490,45 @@ export const FIREBASE_AUTH_DELETE_ACCOUNT = new ApiSettings('/accounts:delete',
+    }
+  })
+  .setResponseValidator((response: BatchDeleteAccountsResponse) => {
+    if (response.errors) {
```
Eliminate the if condition with:

`const errors = response.errors || [];`

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -785,6 +851,62 @@ export abstract class AbstractAuthRequestHandler {
+   *     looked up users. Possibly empty if no users were looked up.
+   */
+  public getAccountInfoByIdentifiers(identifiers: UserIdentifier[]): Promise<object> {
+    if (identifiers.length === 0) {
```
Should probably also do `!validator.isArray(identifiers)` and throw.

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -785,6 +851,62 @@ export abstract class AbstractAuthRequestHandler {
+    const request: GetAccountInfoRequest = {};
+
+    for (const id of identifiers) {
+      if (isUidIdentifier(id)) {
```
I feel like a polymorphic design would have cleaned this clutter up a lot, but such APIs are not very user friendly in JS. How about we at least push some of this logic into small helper functions?

```
if (isUserIdentifier(id)) {
  handleUserIdentifier(id, request);
} else if (isEmailIdentifier(id)) {
  handleEmailIdentifier(id, request);
}
// and so on...
```

hiranya911 commented on src/auth/auth-api-request.ts:
```diff
@@ -882,6 +1004,30 @@ export abstract class AbstractAuthRequestHandler {
   }
 
+  public deleteAccounts(uids: string[], force: boolean): Promise<BatchDeleteAccountsResponse> {
+    if (uids.length === 0) {
```
Check array.

Is the force option needed?

hiranya911 commented on test/unit/auth/auth.spec.ts:
```diff
@@ -1130,6 +1131,84 @@ AUTH_CONFIGS.forEach((testConfig) => {
+      });
+
+      it('should return no results when given no identifiers', () => {
+        return auth.getUsers([])
```
Test for null, undefined and other non-array things.

hiranya911 commented on test/unit/auth/auth.spec.ts:
```diff
@@ -1130,6 +1131,84 @@ AUTH_CONFIGS.forEach((testConfig) => {
+          });
+      });
+
+      it('should be rejected when given an invalid uid', () => {
```
I think we ought to clearly document this behavior. If any of the identifiers are invalid/malformed, it results in a total failure, as opposed to the partial failure that occurs when some identifiers are not available.

hiranya911 commented on test/unit/auth/auth.spec.ts:
```diff
@@ -1130,6 +1131,84 @@ AUTH_CONFIGS.forEach((testConfig) => {
+          .with.property('code', 'auth/invalid-provider-id');
+      });
+
+      it('should be rejected when given a single bad identifier', () => {
```
Also add a happy path test case, where all IDs are valid and all users are present.

rsgowman replied:
Done, though I've combined that with a notFound user. (Let me know if you want that to be explicitly split out.)

hiranya911 commented on test/unit/auth/auth.spec.ts:
```diff
@@ -1207,6 +1286,63 @@ AUTH_CONFIGS.forEach((testConfig) => {
+          .with.property('code', 'auth/invalid-uid');
+      });
+
+      it('should index errors correctly in result', async () => {
```
Add happy path test case

rsgowman replied:
Done.

hiranya911 commented on test/integration/auth.spec.ts:
```diff
@@ -164,6 +164,136 @@ describe('admin.auth', () => {
+    const testUser3 = { uid: 'uid3', email: 'user3@example.com', phoneNumber: '+15555550003' };
+    const usersToCreate = [ testUser1, testUser2, testUser3 ];
+
+    const importUser1: admin.auth.UserImportRecord = {
```
Add a comment why this is here. I assume it's because you wanted a test user with a specific provider config.

rsgowman replied:
Done.

hiranya911 commented on test/integration/auth.spec.ts:
```diff
@@ -164,6 +164,136 @@ describe('admin.auth', () => {
+        .then((getUsersResult) => getUsersResult.users)
+        .then(mapUserRecordsToUidEmailPhones);
+
+      expect(users).to.have.deep.members([testUser1, testUser3]);
```
Should also assert notFound users

rsgowman replied:
Done.

hiranya911 commented on test/integration/auth.spec.ts:
```diff
@@ -1191,6 +1321,63 @@ describe('admin.auth', () => {
+
+      return admin.auth().deleteUsers([uid1, uid2])
+        .then((deleteUsersResult) => {
+          expect(deleteUsersResult.successCount).to.equal(2);
```
This doesn't look right.

rsgowman replied:
Looks ok to me? deleteUsers is idempotent, so deleteUsers(['i_exist', 'but_i_dont']) should result in a success count of 2, i.e. both users no longer exist. Right?

hiranya911 replied:
Ah, idempotency was what I was missing 👍

hiranya911 commented on test/integration/auth.spec.ts:
```diff
@@ -1191,6 +1321,63 @@ describe('admin.auth', () => {
+        });
+    });
+
+    it('is idempotent', async () => {
```
Was this the decided upon behavior?

rsgowman replied:
Yes; see the clarifications section of the api proposal: "DeleteUsers() call that specifies only users that don’t exist (or no users at all) results in an empty result set rather than an error. This allows developers to not need to special case this situation, as well as makes this call idempotent. (The returned SuccessCount value would indicate the number of users that no longer exist, regardless of their initial state, i.e. it would also include users that already did not exist.)"

If you wanted to push back on that, we still could; we haven't fully committed to this behaviour yet. (fwiw, I prefer it as is.) In order to go down that route, we'd have to make a backend change too though, since the backend also behaves this way and doesn't provide any indication as to whether an account was deleted or never existed in the first place.

hiranya911 replied:
I see. My only concern is that it's different from how the existing `deleteUser()` API works. But I also think the behavior proposed here is actually better/convenient for bulk deletes.

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -463,12 +468,21 @@ export const FIREBASE_AUTH_DOWNLOAD_ACCOUNT = new ApiSettings('/accounts:batchGe
-  .setRequestValidator((request: any) => {
-    if (!request.localId && !request.email && !request.phoneNumber) {
+  .setRequestValidator((request: GetAccountInfoRequest) => {
+    if (!request.localId && !request.email && !request.phoneNumber && !request.federatedUserId) {
```
Do you validate the `federatedUserId` content?

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -481,6 +495,21 @@ export const FIREBASE_AUTH_GET_ACCOUNT_INFO = new ApiSettings('/accounts:lookup'
+ * Instantiates the getAccountInfo endpoint settings for use when fetching info
+ * for multiple accounts.
+ */
+export const FIREBASE_AUTH_GET_ACCOUNTS_INFO = new ApiSettings('/accounts:lookup', 'POST')
```
Why do you need to duplicate `FIREBASE_AUTH_GET_ACCOUNT_INFO` above? It is the same underlying API with similar request /response format.

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -492,6 +521,46 @@ export const FIREBASE_AUTH_DELETE_ACCOUNT = new ApiSettings('/accounts:delete',
+
+export const FIREBASE_AUTH_BATCH_DELETE_ACCOUNTS = new ApiSettings('/accounts:batchDelete', 'POST')
+  .setRequestValidator((request: BatchDeleteAccountsRequest) => {
+    if (!request.localIds) {
```
shouldn't you also validate the `force` field?

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -492,6 +521,46 @@ export const FIREBASE_AUTH_DELETE_ACCOUNT = new ApiSettings('/accounts:delete',
+  .setResponseValidator((response: BatchDeleteAccountsResponse) => {
+    const errors = response.errors || [];
+    errors.forEach((batchDeleteErrorInfo) => {
+      if (batchDeleteErrorInfo.index === undefined) {
```
nit: `typeof batchDeleteErrorInfo.index === 'undefined'`

bojeil-google commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,20 @@ declare namespace admin.auth {
+     * @param uids The `uids` corresponding to the users to delete.
+     *
+     * @return A Promise that resolves to the total number of successful/failed
+     *     deletions, as well as the array of errors that correspond to the
```
corresponds

bojeil-google commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,20 @@ declare namespace admin.auth {
     deleteUser(uid: string): Promise<void>;
 
+    /**
+     * Deletes the users specified by the given identifiers.
```
identifiers in this case is always uids.

bojeil-google commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,20 @@ declare namespace admin.auth {
+     *     deletions, as well as the array of errors that correspond to the
+     *     failed deletions.
+     */
+    deleteUsers(uids: string[]): Promise<admin.auth.DeleteUsersResult>;
```
I am a bit confused. Why is `forceDelete` not included? This was approved by the API council in https://docs.google.com/document/d/1mG0ilmh8H-C6DYXA9oOQUro2MP0on_9k7Qf4oCLRyBk/

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -725,6 +794,44 @@ export abstract class AbstractAuthRequestHandler {
+
+  private static addProviderToRequest(id: ProviderIdentifier, request: GetAccountInfoRequest): GetAccountInfoRequest {
+    if (!validator.isNonEmptyString(id.providerUid) || !validator.isNonEmptyString(id.providerId)) {
+      throw new FirebaseAuthError(AuthClientErrorCode.INVALID_PROVIDER_ID);
```
I think invalid provider ID is different that invalid provider uid. The error is confusing here.

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -811,6 +918,44 @@ export abstract class AbstractAuthRequestHandler {
+  public getAccountInfoByIdentifiers(identifiers: UserIdentifier[]): Promise<object> {
+    if (identifiers.length === 0) {
+      return Promise.resolve({users: []});
+    } else if (identifiers.length > 100) {
```
Maybe define 100 as a constant like we do for other similar max fields?

bojeil-google commented on src/auth/auth-api-request.ts:
```diff
@@ -908,6 +1053,30 @@ export abstract class AbstractAuthRequestHandler {
+  public deleteAccounts(uids: string[], force: boolean): Promise<BatchDeleteAccountsResponse> {
+    if (uids.length === 0) {
+      return Promise.resolve({});
+    } else if (uids.length > 100) {
```
Same here regarding defining a const for this upper limit.

bojeil-google commented on src/auth/auth.ts:
```diff
@@ -200,6 +225,61 @@ export class BaseAuth<T extends AbstractAuthRequestHandler> {
+         * Checks if the specified identifier is within the list of
+         * UserRecords.
+         */
+        const isUserFound = ((id: UserIdentifier, urs: UserRecord[]): boolean => {
```
what does `urs` stand for? Maybe pick a better name?

hiranya911 commented on src/auth/auth.ts:
```diff
@@ -271,6 +351,52 @@ export class BaseAuth<T extends AbstractAuthRequestHandler> {
+          const errMsgToError = (msg?: string): FirebaseAuthError => {
+            // We unconditionally set force=true, so the 'NOT_DISABLED' error
+            // should not be possible.
+            if (msg && msg.startsWith('NOT_DISABLED :')) {
```
Nit:

```
const code = msg && msg.startsWith('NOT_DISABLED') ? 
  AuthClientErrorCode.USER_NOT_DISABLED : AuthClientErrorCode.INTERNAL_ERROR;
return new FirebaseAuthError(code, ...);
```

rsgowman replied:
Done.

hiranya911 commented on src/index.d.ts:
```diff
@@ -1529,6 +1629,23 @@ declare namespace admin.auth {
+     * There are no ordering guarantees; in particular, the nth entry in the result list is not
+     * guaranteed to correspond to the nth entry in the input parameters list.
+     *
+     * Only a maximum of 100 identifiers may be supplied. If more than 100 identifiers are supplied,
```
I think this method rejects when array.length > 100. Let's verify and update the comment.

rsgowman replied:
i.e. throw vs failed promise? It should throw. (There's a test to ensure it here: test/unit/auth/auth.spec.ts:1142.)

hiranya911 replied:
Ah, I see now. Thanks. The terminology used in the test case descriptions could need some work though.

hiranya911 commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,20 @@ declare namespace admin.auth {
+    /**
+     * Deletes the users specified by the given uids.
+     *
+     * Deletes multiple existing users. Deleting a non-existing user won't
```
Please document array length constraints.

rsgowman replied:
Oops; I hadn't updated this based on the design doc. I've added the full docs now. (Which includes the 1000 thing, but also touches on the rate limitation.)

hiranya911 commented on test/unit/auth/auth.spec.ts:
```diff
@@ -1130,6 +1131,132 @@ AUTH_CONFIGS.forEach((testConfig) => {
+        stubs = [];
+      });
+
+      it('should be rejected when given more than 100 identifiers', () => {
```
Nit: `should throw`

rsgowman replied:
Done.

egilmorez commented on src/index.d.ts:
```diff
@@ -921,6 +940,32 @@ declare namespace admin.auth {
+
+    /**
+     * The number of users that were deleted successfully (possibly zero).
+     * Users that did not exist prior to calling deleteUsers() will be
```
Suggest "are considered"

rsgowman replied:
Done.

egilmorez commented on src/index.d.ts:
```diff
@@ -921,6 +940,32 @@ declare namespace admin.auth {
+    successCount: number;
+
+    /**
+     * A list of FirebaseArrayIndexError instances describing the errors that
```
Suggest backticks.

rsgowman replied:
Done.

egilmorez commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,31 @@ declare namespace admin.auth {
+     * Deletes the users specified by the given uids.
+     *
+     * Deleting a non-existing user won't generate an error. (i.e. this method
+     * is idempotent.) Non-existing users will be considered to be successfully
```
Suggest "are considered . . . are counted"

rsgowman replied:
Done.

egilmorez commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,31 @@ declare namespace admin.auth {
+     * Deleting a non-existing user won't generate an error. (i.e. this method
+     * is idempotent.) Non-existing users will be considered to be successfully
+     * deleted, and will therefore be counted in the
+     * DeleteUsersResult.successCount value.
```
Suggest backticks.

egilmorez commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,31 @@ declare namespace admin.auth {
+     * DeleteUsersResult.successCount value.
+     *
+     * Only a maximum of 1000 identifiers may be supplied. If more than 1000
+     * identifiers are supplied, this method will immediately throw a
```
Suggest "method throws"

rsgowman replied:
Done.

egilmorez commented on src/index.d.ts:
```diff
@@ -1529,6 +1640,23 @@ declare namespace admin.auth {
+     * guaranteed to correspond to the nth entry in the input parameters list.
+     *
+     * Only a maximum of 100 identifiers may be supplied. If more than 100 identifiers are supplied,
+     * this method will immediately throw a FirebaseAuthError.
```
Suggest "method throws"

rsgowman replied:
Done.

RelaxedPear commented on test/integration/auth.spec.ts:
```diff
@@ -168,6 +168,138 @@ describe('admin.auth', () => {
+        // New users should not have a lastRefreshTime set.
+        expect(newUserRecord.metadata.lastRefreshTime).to.be.null;
+
+        // Login to cause the lastRefreshTime to be set.
```
Suggest:
"// Log in to set the lastRefreshTime."

rsgowman replied:
Done.

RelaxedPear commented on src/index.d.ts:
```diff
@@ -492,10 +492,16 @@ declare namespace admin.auth {
     creationTime: string;
 
+    /**
+     * The time at which the user was last active (ID token refreshed), or null
```
"The time at which the user was last active (ID token refreshed), formatted as a UTC Date string (e.g., 'Sat, 03 Feb 2001 04:05:06 GMT'). Returns null if the user was never active."

rsgowman replied:
Done.

RelaxedPear commented on src/index.d.ts:
```diff
@@ -1487,6 +1573,30 @@ declare namespace admin.auth {
+    /**
+     * Deletes the users specified by the given uids.
+     *
+     * Deleting a non-existing user won't generate an error. (i.e. this method
```
Deleting a non-existing user won't generate an error (i.e. this method is idempotent).

rsgowman replied:
Done.

bojeil-google commented on src/index.d.ts:
```diff
@@ -604,7 +604,7 @@ declare namespace admin.credential {
   interface Credential {
 
-    /**
+   /**
```
Comment alignment seem to be off here.

bojeil-google commented:
Granted coverage is still adequate but the testing format does not align with the existing pattern used here. Personally I prefer sticking with the current pattern.

The pattern has been to test the RPC logic in auth-api-request.spec.ts
Then test the adequate auth-api-request calls are made in auth.spec.ts.

rsgowman commented:
> Granted coverage is still adequate but the testing format does not align with the existing pattern used here. Personally I prefer sticking with the current pattern.
> 
> The pattern has been to test the RPC logic in auth-api-request.spec.ts
> Then test the adequate auth-api-request calls are made in auth.spec.ts.

Mostly done. I've moved some tests over entirely (where possible), kept some tests in auth.spec.ts (when those tests test behaviour entirely confined to auth.ts) and split some between the two (as auth.ts modifies the results from auth-api-request.ts, most notably adding a notFound parameter to the results as well as parsing everything in to a UserRecord.)

rsgowman commented:
> LG, thanks!
> 
> There's an instance of "deleteUsers()" in line is 680 of auth.d.ts that should be backticked, but we can get that later if it's a PITA.

Not a problem; fixed.

bojeil-google commented on test/unit/auth/auth-api-request.spec.ts:
```diff
@@ -1727,6 +1920,35 @@ AUTH_REQUEST_HANDLER_TESTS.forEach((handler) => {
       });
     });
 
+    describe('deleteAccounts', () => {
```
Can you add tests for rpc logic for this API?

bojeil-google commented on test/unit/auth/auth-api-request.spec.ts:
```diff
@@ -1947,6 +1950,20 @@ AUTH_REQUEST_HANDLER_TESTS.forEach((handler) => {
           .with.property('code', 'auth/invalid-uid');
       });
+
+      it('should be fulfilled given a valids uids', async () => {
```
given valid
